// angular-maps/esm5/src/services/infobox.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * This class defines the contract for an InfoBoxService. Each Map Architecture provider is expected the furnish a concrete implementation.
 *
 * @export
 * @abstract
 * @abstract
 */
var InfoBoxService = /** @class */ (function () {
    function InfoBoxService() {
    }
    InfoBoxService.decorators = [
        { type: Injectable },
    ];
    return InfoBoxService;
}());
export { InfoBoxService };
function InfoBoxService_tsickle_Closure_declarations() {
    /**
     * Adds an info window to the map or layer.
     *
     * @abstract
     * \@memberof InfoBoxService
     * @abstract
     * @param {?} info - {\@link InfoBoxComponent} component object representing the infobox.
     *
     * @return {?}
     */
    InfoBoxService.prototype.AddInfoWindow = function (info) { };
    /**
     * Closes an infobox that is open.
     *
     * @abstract
     * \@memberof InfoBoxService
     * @abstract
     * @param {?} info - {\@link InfoBoxComponent} component object representing the infobox.
     * @return {?} - A promise that is fullfilled when the infobox has been closed.
     *
     */
    InfoBoxService.prototype.Close = function (info) { };
    /**
     * Subscribe to events on the infowindow.
     *
     * @abstract
     * \@memberof InfoBoxService
     * @abstract
     * @template T
     * @param {?} event
     * @param {?} infoBoxComponent
     * @return {?} - Observable emiting an instance of T each time the event occurs.
     *
     */
    InfoBoxService.prototype.CreateEventObservable = function (event, infoBoxComponent) { };
    /**
     * Deletes an infobox.
     *
     * @abstract
     * \@memberof InfoBoxService
     * @abstract
     * @param {?} info - {\@link InfoBoxComponent} component object representing the infobox.
     * @return {?} - A promise that is fullfilled when the infobox has been deleted.
     *
     */
    InfoBoxService.prototype.DeleteInfoWindow = function (info) { };
    /**
     * Opens an infobox that is closed.
     *
     * @abstract
     * \@memberof InfoBoxService
     * @abstract
     * @param {?} info - {\@link InfoBoxComponent} component object representing the infobox.
     * @param {?=} loc
     * @return {?} - A promise that is fullfilled when the infobox has been opened.
     *
     */
    InfoBoxService.prototype.Open = function (info, loc) { };
    /**
     * Sets the infobox options.
     *
     * @abstract
     * \@memberof InfoBoxService
     * @abstract
     * @param {?} info - {\@link InfoBoxComponent} component object representing the infobox.
     * @param {?} options - {\@link IInfoWindowOptions} object containing the options to set. Options provided are
     * merged with the existing options of the underlying infobox.
     * @return {?} - A promise that is fullfilled when the infobox options have been updated.
     *
     */
    InfoBoxService.prototype.SetOptions = function (info, options) { };
    /**
     * Set the position of the infobox based on the properties set on the InfoBox component.
     *
     * @abstract
     * \@memberof InfoBoxService
     * @abstract
     * @param {?} info - {\@link InfoBoxComponent} component object representing the infobox.
     * @param {?=} latlng - The position to set
     * @return {?} - A promise that is fullfilled when the infobox position has been updated.
     *
     */
    InfoBoxService.prototype.SetPosition = function (info, latlng) { };
}
