// angular-maps/esm5/src/services/map.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * Abstract class to implement map api. A concrete implementation should be created for each
 * Map provider supported (e.g. Bing, Goolge, ESRI)
 *
 * @export
 * @abstract
 * @abstract
 */
var MapService = /** @class */ (function () {
    function MapService() {
    }
    /**
     * Gets a random geo locations filling the bounding box.
     *
     * \@memberof MapService
     * @param {?} count - number of locations to return
     * @param {?} bounds  - bounding box.
     * @return {?} - Array of geo locations.
     */
    MapService.GetRandonLocations = /**
     * Gets a random geo locations filling the bounding box.
     *
     * \@memberof MapService
     * @param {?} count - number of locations to return
     * @param {?} bounds  - bounding box.
     * @return {?} - Array of geo locations.
     */
    function (count, bounds) {
        var /** @type {?} */ a = [];
        var /** @type {?} */ _getRandomLocation = function (b) {
            var /** @type {?} */ lat = Math.random() * (b.maxLatitude - b.minLatitude) + b.minLatitude;
            var /** @type {?} */ lng = 0;
            if (crossesDateLine) {
                lng = Math.random() * (b.minLongitude + 360 - b.maxLongitude) + b.maxLongitude;
                if (lng > 180) {
                    lng = lng - 360;
                }
            }
            else {
                lng = Math.random() * (b.maxLongitude - b.minLongitude) + b.minLongitude;
            }
            var /** @type {?} */ p = { latitude: lat, longitude: lng };
            return p;
        };
        var /** @type {?} */ crossesDateLine = false;
        if (bounds == null) {
            bounds = /** @type {?} */ ({
                maxLatitude: 360,
                minLatitude: 0,
                maxLongitude: 170,
                minLongitude: 0
            });
        }
        if (bounds.center.longitude < bounds.minLongitude || bounds.center.longitude > bounds.maxLongitude) {
            crossesDateLine = true;
        }
        if (!count || count <= 0) {
            return [_getRandomLocation(bounds)];
        }
        for (var /** @type {?} */ r = 0; r < count; r++) {
            a.push(_getRandomLocation(bounds));
        }
        return a;
    };
    MapService.decorators = [
        { type: Injectable },
    ];
    return MapService;
}());
export { MapService };
function MapService_tsickle_Closure_declarations() {
    /**
     * Gets the Map control instance underlying the implementation
     *
     * \@readonly
     * \@memberof MapService
     * @abstract
     * @return {?}
     */
    MapService.prototype.MapInstance = function () { };
    /**
     * Gets a Promise for a Map control instance underlying the implementation. Use this instead of {\@link MapInstance} if you
     * are not sure if and when the instance will be created.
     * \@readonly
     * \@memberof MapService
     * @abstract
     * @return {?}
     */
    MapService.prototype.MapPromise = function () { };
    /**
     * Gets the maps physical size.
     *
     * \@readonly
     * @abstract
     * \@memberof MapService
     * @abstract
     * @return {?}
     */
    MapService.prototype.MapSize = function () { };
    /**
     * Creates a canvas overlay layer to perform custom drawing over the map with out
     * some of the overhead associated with going through the Map objects.
     * \@memberof MapService
     * @abstract
     * @abstract
     * @param {?} drawCallback A callback function that is triggered when the canvas is ready to be
     * rendered for the current map view.
     * @return {?} - Promise of a {\@link CanvasOverlay} object.
     */
    MapService.prototype.CreateCanvasOverlay = function (drawCallback) { };
    /**
     * Creates a map cluster layer within the map context
     *
     * \@memberof MapService
     * @abstract
     * @param {?} options - Options for the layer. See {\@link IClusterOptions}.
     * @return {?} - Promise of a {\@link Layer} object, which models the underlying native layer object.
     *
     */
    MapService.prototype.CreateClusterLayer = function (options) { };
    /**
     * Creates an information window for a map position
     *
     * \@memberof MapService
     * @abstract
     * @param {?=} options
     * @return {?} - Promise of a {\@link InfoWindow} object, which models the underlying natvie infobox object.
     *
     */
    MapService.prototype.CreateInfoWindow = function (options) { };
    /**
     * Creates a map layer within the map context
     *
     * \@memberof MapService
     * @abstract
     * @param {?} options - Options for the layer. See {\@link ILayerOptions}
     * @return {?} - Promise of a {\@link Layer} object, which models the underlying native layer object.
     *
     */
    MapService.prototype.CreateLayer = function (options) { };
    /**
     * Creates a map instance
     *
     * \@memberof MapService
     * @abstract
     * @param {?} el - HTML element to host the map.
     * @param {?} mapOptions - Map options
     * @return {?} - Promise fullfilled once the map has been created.
     *
     */
    MapService.prototype.CreateMap = function (el, mapOptions) { };
    /**
     * Creates a map marker within the map context
     *
     * \@memberof MapService
     * @abstract
     * @param {?} options
     * @return {?} - Promise of a {\@link Marker} object, which models the underlying native pushpin object.
     *
     */
    MapService.prototype.CreateMarker = function (options) { };
    /**
     * Creates a polygon within the map context
     *
     * @abstract
     * \@memberof MapService
     * @abstract
     * @param {?} options - Options for the polygon. See {\@link IPolygonOptions}.
     * @return {?} - Promise of a {\@link Polygon} object, which models the underlying native polygon.
     *
     */
    MapService.prototype.CreatePolygon = function (options) { };
    /**
     * Creates a polyline within the map context
     *
     * @abstract
     * \@memberof MapService
     * @abstract
     * @param {?} options - Options for the polyline. See {\@link IPolylineOptions}.
     * @return {?} - Promise of a {\@link Polyline} object (or an array thereof for complex paths),
     * which models the underlying native polyline.
     *
     */
    MapService.prototype.CreatePolyline = function (options) { };
    /**
     * Deletes a layer from the map.
     *
     * \@memberof MapService
     * @abstract
     * @param {?} layer - Layer to delete. See {\@link Layer}.
     * @return {?} - Promise fullfilled when the layer has been removed.
     *
     */
    MapService.prototype.DeleteLayer = function (layer) { };
    /**
     * Dispaose the map and associated resoures.
     *
     * \@memberof MapService
     * @abstract
     * @return {?}
     */
    MapService.prototype.DisposeMap = function () { };
    /**
     * Gets the geo coordinates of the map bounds
     *
     * \@memberof MapService
     * @abstract
     * @return {?} - A promise that when fullfilled contains the bounding box of the screen. See {\@link IBox}.
     *
     */
    MapService.prototype.GetBounds = function () { };
    /**
     * Gets the geo coordinates of the map center
     *
     * \@memberof MapService
     * @abstract
     * @return {?} - A promise that when fullfilled contains the goe location of the center. See {\@link ILatLong}.
     *
     */
    MapService.prototype.GetCenter = function () { };
    /**
     * Gets the current zoom level of the map.
     *
     * \@memberof MapService
     * @abstract
     * @return {?} - A promise that when fullfilled contains the zoom level.
     *
     */
    MapService.prototype.GetZoom = function () { };
    /**
     * Provides a conversion of geo coordinates to pixels on the map control.
     *
     * \@memberof MapService
     * @abstract
     * @param {?} loc - The geo coordinates to translate.
     * @return {?} - Promise of an {\@link IPoint} interface representing the pixels. This promise resolves to null
     * if the goe coordinates are not in the view port.
     *
     */
    MapService.prototype.LocationToPoint = function (loc) { };
    /**
     * Provides a conversion of geo coordinates to pixels on the map control.
     *
     * \@memberof MapService
     * @abstract
     * @param {?} locs
     * @return {?} - Promise of an {\@link IPoint} interface array representing the pixels.
     *
     */
    MapService.prototype.LocationsToPoints = function (locs) { };
    /**
     * Centers the map on a geo location.
     *
     * \@memberof MapService
     * @abstract
     * @param {?} latLng - GeoCoordinates around which to center the map. See {\@link ILatLong}
     * @return {?} - Promise that is fullfilled when the center operations has been completed.
     *
     */
    MapService.prototype.SetCenter = function (latLng) { };
    /**
     * Sets the generic map options.
     *
     * \@memberof MapService
     * @abstract
     * @param {?} options - Options to set.
     *
     * @return {?}
     */
    MapService.prototype.SetMapOptions = function (options) { };
    /**
     * Sets the view options of the map.
     *
     * \@memberof MapService
     * @abstract
     * @param {?} options - Options to set.
     *
     * @return {?}
     */
    MapService.prototype.SetViewOptions = function (options) { };
    /**
     * Sets the zoom level of the map.
     *
     * \@memberof MapService
     * @abstract
     * @param {?} zoom - Zoom level to set.
     * @return {?} - A Promise that is fullfilled once the zoom operation is complete.
     *
     */
    MapService.prototype.SetZoom = function (zoom) { };
    /**
     * Creates an event subscription
     *
     * \@memberof MapService
     * @abstract
     * @template E
     * @param {?} eventName - The name of the event (e.g. 'click')
     * @return {?} - An observable of tpye E that fires when the event occurs.
     *
     */
    MapService.prototype.SubscribeToMapEvent = function (eventName) { };
    /**
     * Triggers the given event name on the map instance.
     *
     * \@memberof MapService
     * @abstract
     * @param {?} eventName - Event to trigger.
     * @return {?} - A promise that is fullfilled once the event is triggered.
     *
     */
    MapService.prototype.TriggerMapEvent = function (eventName) { };
}
