// angular-maps/esm5/src/services/polyline.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * The abstract class represents the contract defintions for a polyline service to be implemented by an acutaly underlying
 * map architecture.
 *
 * @export
 * @abstract
 * @abstract
 */
var PolylineService = /** @class */ (function () {
    function PolylineService() {
    }
    PolylineService.decorators = [
        { type: Injectable },
    ];
    return PolylineService;
}());
export { PolylineService };
function PolylineService_tsickle_Closure_declarations() {
    /**
     * Adds a polyline to a map. Depending on the polyline context, the polyline will either by added to the map or a
     * correcsponding layer.
     *
     * @abstract
     * \@memberof PolylineService
     * @abstract
     * @param {?} polyline - The {\@link MapPolylineDirective} to be added.
     *
     * @return {?}
     */
    PolylineService.prototype.AddPolyline = function (polyline) { };
    /**
     * Registers an event delegate for a marker.
     *
     * @abstract
     * \@memberof PolylineService
     * @abstract
     * @template T
     * @param {?} eventName - The name of the event to register (e.g. 'click')
     * @param {?} polyline - The {\@link MapPolylineDirective} for which to register the event.
     * @return {?} - Observable emiting an instance of T each time the event occurs.
     *
     */
    PolylineService.prototype.CreateEventObservable = function (eventName, polyline) { };
    /**
     * Deletes a polyline.
     *
     * @abstract
     * \@memberof PolylineService
     * @abstract
     * @param {?} polyline - {\@link MapPolylineDirective} to be deleted.
     * @return {?} - A promise fullfilled once the polyline has been deleted.
     *
     */
    PolylineService.prototype.DeletePolyline = function (polyline) { };
    /**
     * Obtains geo coordinates for the marker on the click location
     *
     * @abstract
     * \@memberof MarkerService
     * @abstract
     * @param {?} e - The mouse event.
     * @return {?} - {\@link ILatLong} containing the geo coordinates of the clicked marker.
     *
     */
    PolylineService.prototype.GetCoordinatesFromClick = function (e) { };
    /**
     * Obtains the polyline model for the polyline allowing access to native implementation functionatiliy.
     *
     * @abstract
     * \@memberof PolylineService
     * @abstract
     * @param {?} polyline - The {\@link MapPolylineDirective} for which to obtain the polyline model.
     * @return {?} - A promise that when fullfilled contains the {\@link Polyline} implementation (or an
     * array of polylines) for complex paths of the underlying platform.
     *
     */
    PolylineService.prototype.GetNativePolyline = function (polyline) { };
    /**
     * Set the polyline options.
     *
     * @abstract
     * \@memberof PolylineService
     * @abstract
     * @param {?} polyline - {\@link MapPolylineDirective} to be updated.
     * @param {?} options - {\@link IPolylineOptions} object containing the options. Options will be merged with the
     * options already on the underlying object.
     * @return {?} - A promise fullfilled once the polyline options have been set.
     *
     */
    PolylineService.prototype.SetOptions = function (polyline, options) { };
    /**
     * Updates the Polyline path
     *
     * @abstract
     * \@memberof PolylineService
     * @abstract
     * @param {?} polyline - {\@link MapPolylineDirective} to be updated.
     * @return {?} - A promise fullfilled once the polyline has been updated.
     *
     */
    PolylineService.prototype.UpdatePolyline = function (polyline) { };
}
