// angular-maps/esm2015/src/services/polygon.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * The abstract class represents the contract defintions for a polygon service to be implemented by an acutaly underlying
 * map architecture.
 *
 * @export
 * @abstract
 * @abstract
 */
export class PolygonService {
}
PolygonService.decorators = [
    { type: Injectable },
];
function PolygonService_tsickle_Closure_declarations() {
    /**
     * Adds a polygon to a map. Depending on the polygon context, the polygon will either by added to the map or a
     * correcsponding layer.
     *
     * @abstract
     * \@memberof PolygonService
     * @abstract
     * @param {?} polygon - The {\@link MapPolygonDirective} to be added.
     *
     * @return {?}
     */
    PolygonService.prototype.AddPolygon = function (polygon) { };
    /**
     * Registers an event delegate for a marker.
     *
     * @abstract
     * \@memberof PolygonService
     * @abstract
     * @template T
     * @param {?} eventName - The name of the event to register (e.g. 'click')
     * @param {?} polygon - The {\@link MapPolygonDirective} for which to register the event.
     * @return {?} - Observable emiting an instance of T each time the event occurs.
     *
     */
    PolygonService.prototype.CreateEventObservable = function (eventName, polygon) { };
    /**
     * Deletes a polygon.
     *
     * @abstract
     * \@memberof PolygonService
     * @abstract
     * @param {?} polygon - {\@link MapPolygonDirective} to be deleted.
     * @return {?} - A promise fullfilled once the polygon has been deleted.
     *
     */
    PolygonService.prototype.DeletePolygon = function (polygon) { };
    /**
     * Obtains geo coordinates for the marker on the click location
     *
     * @abstract
     * \@memberof MarkerService
     * @abstract
     * @param {?} e - The mouse event.
     * @return {?} - {\@link ILatLong} containing the geo coordinates of the clicked marker.
     *
     */
    PolygonService.prototype.GetCoordinatesFromClick = function (e) { };
    /**
     * Obtains the polygon model for the polygon allowing access to native implementation functionatiliy.
     *
     * @abstract
     * \@memberof PolygonService
     * @abstract
     * @param {?} polygon - The {\@link MapPolygonDirective} for which to obtain the polygon model.
     * @return {?} - A promise that when fullfilled contains the {\@link Polygon} implementation of the underlying platform.
     *
     */
    PolygonService.prototype.GetNativePolygon = function (polygon) { };
    /**
     * Set the polygon options.
     *
     * @abstract
     * \@memberof PolygonService
     * @abstract
     * @param {?} polygon - {\@link MapPolygonDirective} to be updated.
     * @param {?} options - {\@link IPolygonOptions} object containing the options. Options will be merged with the
     * options already on the underlying object.
     * @return {?} - A promise fullfilled once the polygon options have been set.
     *
     */
    PolygonService.prototype.SetOptions = function (polygon, options) { };
    /**
     * Updates the Polygon path
     *
     * @abstract
     * \@memberof PolygonService
     * @abstract
     * @param {?} polygon - {\@link MapPolygonDirective} to be updated.
     * @return {?} - A promise fullfilled once the polygon has been updated.
     *
     */
    PolygonService.prototype.UpdatePolygon = function (polygon) { };
}
